test: add cli test to validate multiple manifest generation paths result in comparable output

btlghrants opened the pull request:
## Description

Adds a new `npm run test:int` (int == "integration") script to invoke a new style of test -- integration tests -- where "integration" means something like "wider in scope than unit tests" but "not requiring a full K8s cluster" (general idea is to have a nice place for non-k8s-required tests to land (e.g. client-side `pepr` CLI commands, like init / build).

## Related Issue

Fixes #1654 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging
- [x] Unit, [Journey](https://github.com/defenseunicorns/pepr/tree/main/journey), [E2E Tests](https://github.com/defenseunicorns/pepr-excellent-examples), [docs](https://github.com/defenseunicorns/pepr/tree/main/docs), [adr](https://github.com/defenseunicorns/pepr/tree/main/adr) added or updated as needed
- [x] [Contributor Guide Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request) followed

btlghrants committed:
wip: initial

btlghrants committed:
wip: update eslint config

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: refactor f/ eslint warnings

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: somethings changing root package.json, ugh, so I'm changing approach

btlghrants committed:
wip: new approach (with NPM_CONFIG_CACHE) working much better

btlghrants committed:
wip: saving proress

btlghrants committed:
wip: saving proress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: saving progress

btlghrants committed:
wip: parallelize them tests!

btlghrants committed:
wip: merge main

btlghrants committed:
wip: rm redundant CI test file

samayer12 commented on integration/cli/build.defaults.test.ts:
```diff
@@ -0,0 +1,52 @@
+
+  describe("builds a module", () => {
+    const id = FILE.split(".").at(1);
+    const mod = `${workdir.path()}/${id}`;
```
```suggestion
    const module = `${workdir.path()}/${id}`;
```

Avoiding abbreviations for short words is a personal style preference of mine

btlghrants replied:
A reasonable preference -- and one I agree with, generally -- when the variable is going to be exported / made part of a public API / used widely / live a long time, but... none of that really applies here.

_Not_ using reasonable shortenings just adds length to the code & makes it harder to scan, IMO, especially for the names of a local variables that are instantiated & die no more than 20 lines later.  But I don't feel that strongly about this case so sure, we can change it.  ¯\\_(ツ)_/¯

On this particular item though: using the name "module" would shadow a node built-in that I'd prefer to leave alone, so... I'll use "testModule" instead.  That work?

samayer12 replied:
`mod` -> `testModule` sounds good, thanks!

samayer12 commented on integration/cli/build.defaults.test.ts:
```diff
@@ -0,0 +1,52 @@
+    const mod = `${workdir.path()}/${id}`;
+
+    beforeAll(async () => {
+      await fs.rm(mod, { recursive: true, force: true });
```
```suggestion
      await fs.rm(module, { recursive: true, force: true });
```

samayer12 commented on integration/cli/build.defaults.test.ts:
```diff
@@ -0,0 +1,52 @@
+        "--skip-post-init",
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
```
```suggestion
      await pepr.tgzifyModule(module);
```

samayer12 commented on integration/cli/build.defaults.test.ts:
```diff
@@ -0,0 +1,52 @@
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
+      await pepr.cli(mod, { cmd: `npm install` });
```
```suggestion
      await pepr.cli(module, { cmd: `npm install` });
```

samayer12 commented on integration/cli/build.defaults.test.ts:
```diff
@@ -0,0 +1,52 @@
+      it(
+        "builds",
+        async () => {
+          const build = await pepr.cli(mod, { cmd: `pepr build` });
```
```suggestion
          const build = await pepr.cli(module, { cmd: `pepr build` });
```

samayer12 commented on integration/cli/build.help.test.ts:
```diff
@@ -0,0 +1,31 @@
+      const res = await pepr.cli(workdir.path(), { cmd: `pepr build ${argz}` });
+      expect(res.exitcode).toBe(0);
+      expect(res.stderr.join("").trim()).toBe("");
+      expect(res.stdout.at(0)).toMatch("Usage: pepr build");
```
```suggestion
      const result = await pepr.cli(workdir.path(), { cmd: `pepr build ${argz}` });
      expect(result.exitcode).toBe(0);
      expect(result.stderr.join("").trim()).toBe("");
      expect(result.stdout.at(0)).toMatch("Usage: pepr build");
```

Avoiding abbreviations for short works is a personal-preference of mine

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+          // TODO: team talk
+          // Should this be writing to stderr? Even with a 0 exit code..?
+          expect(build.stderr.join("").trim()).toContain("Error: Cannot find module");
+          // TODO: end
```
Let's remove the `TODO` comment and show this as an issue in the backlog

btlghrants replied:
The plan was to do this _after_ the PR merged and to leave these comments in as a to-be-removed part of those Issues' "Definition of Done".

That would leave _this_ test as fully descriptive of the current behavior _and_ leave a searchable pin for where the new tests should live once work starts on follow-up issues but if I'm the only one that finds that sort of thing valuable then I can remove it.

To be clear though, what's your objection to leaving this nugget in a "TODO" comment?

samayer12 replied:
Sure, that works for me. At present we don't have an automated check that warns on lingering-TODOs (though, we could...). In my experience, they have a nasty habit of accumulating over time.

As long as we don't forget to revisit this topic, I'm cool with leaving it in 👍

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
+      await pepr.cli(mod, { cmd: `npm install` });
```
```suggestion
      await pepr.tgzifyModule(module);
      await pepr.cli(module, { cmd: `npm install` });
```

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+    const mod = `${workdir.path()}/${id}`;
+
+    beforeAll(async () => {
+      await fs.rm(mod, { recursive: true, force: true });
```
```suggestion
      await fs.rm(module, { recursive: true, force: true });
```

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+        "builds",
+        async () => {
+          const argz = [`--no-embed`].join(" ");
+          const build = await pepr.cli(mod, { cmd: `pepr build ${argz}` });
```
```suggestion
          const build = await pepr.cli(module, { cmd: `pepr build ${argz}` });
```

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+
+          expect(build.stdout.join("").trim()).toContain("");
+
+          packageJson = await resource.oneFromFile(`${mod}/package.json`);
```
```suggestion
          packageJson = await resource.oneFromFile(`${module}/package.json`);
```

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+            `${mod}/dist/pepr-${uuid}.js.LEGAL.txt`,
+            `${mod}/dist/pepr-module-${uuid}.yaml`,
+            `${mod}/dist/zarf.yaml`,
+            `${mod}/dist/${uuid}-chart/`,
```
```suggestion
            `${module}/dist/pepr-${uuid}.js`,
            `${module}/dist/pepr-${uuid}.js.map`,
            `${module}/dist/pepr-${uuid}.js.LEGAL.txt`,
            `${module}/dist/pepr-module-${uuid}.yaml`,
            `${module}/dist/zarf.yaml`,
            `${module}/dist/${uuid}-chart/`,
```

samayer12 commented on integration/cli/build.noembed.test.ts:
```diff
@@ -0,0 +1,93 @@
+          const found = [
+            `${mod}/dist/pepr.js`,
+            `${mod}/dist/pepr.js.map`,
+            `${mod}/dist/pepr.js.LEGAL.txt`,
```
```suggestion
            `${module}/dist/pepr.js`,
            `${module}/dist/pepr.js.map`,
            `${module}/dist/pepr.js.LEGAL.txt`,
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+
+  describe("builds a module", () => {
+    const id = FILE.split(".").at(1);
+    const mod = `${workdir.path()}/${id}`;
```
```suggestion
    const module = `${workdir.path()}/${id}`;
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+    const mod = `${workdir.path()}/${id}`;
+
+    beforeAll(async () => {
+      await fs.rm(mod, { recursive: true, force: true });
```
```suggestion
      await fs.rm(module, { recursive: true, force: true });
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+        "--skip-post-init",
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
```
```suggestion
      await pepr.tgzifyModule(module);
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
+      await pepr.cli(mod, { cmd: `npm install` });
```
```suggestion
      await pepr.cli(module, { cmd: `npm install` });
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+    describe("using non-conflicting build override options", () => {
+      const entryPoint = "pepr2.ts";
+      const customImage = "pepr:override";
+      const outputDir = `${mod}/out`;
```
```suggestion
      const outputDir = `${module}/out`;
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+      it(
+        "builds",
+        async () => {
+          await fs.rename(`${mod}/pepr.ts`, `${mod}/${entryPoint}`);
```
```suggestion
          await fs.rename(`${module}/pepr.ts`, `${module}/${entryPoint}`);
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+            `--withPullSecret ${withPullSecret}`,
+            `--zarf ${zarf}`,
+          ].join(" ");
+          const build = await pepr.cli(mod, { cmd: `pepr build ${argz}` });
```
```suggestion
          const build = await pepr.cli(module, { cmd: `pepr build ${argz}` });
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+          expect(build.stderr.join("").trim()).toBe("");
+          expect(build.stdout.join("").trim()).toContain("K8s resource for the module saved");
+
+          packageJson = await resource.oneFromFile(`${mod}/package.json`);
```
```suggestion
          packageJson = await resource.oneFromFile(`${module}/package.json`);
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+      );
+
+      const getDepConImg = (deploy: kind.Deployment, container: string): string => {
+        return deploy!.spec!.template!.spec!.containers.filter(f => f.name === container).at(0)!
```
```suggestion
        return deploy!.spec!.template!.spec!.containers.filter(file => file.name === container).at(0)!
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+          .image!;
+      };
+
+      it("--entry-point, works", async () => {
+        // build would fail if given entrypoint didn't exist, so... no-op test!
+      });
```
Is this no-op test here just to enumerate ways to use the CLI tool? Could we have it actually do something?

btlghrants replied:
That no-op test is there just so that a reader of the tests can see from the test output that the `--entry-point` flag _is not_ being forgotten -- simply "successfully building" is the success bar here.

I guess there probably _are_ things that could be inserted there but I don't think they'd be anything more than busywork / test theater because:
- all entrypoint flag does is change the name of a file being looked for during build, and
- we establish / setup / name those files ourselves as part of test setup (i.e. moving `pepr.ts` to `pepr2.ts` and then telling the entrypoint to look for `pepr2.ts`).

Everything else from that point forward is bog standard.  The output module produced is the same, whatever the entrypoint (so long as it works to build).

Open to suggestions if you've got 'em, but I couldn't think of anything to test there that'd add anything more than extra runtime.

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+      });
+
+      it("--output-dir, works", async () => {
+        const dist = `${mod}/dist`;
```
```suggestion
        const dist = `${module}/dist`;
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+        expect(await file.exists(dist)).toBe(false);
+
+        const out = outputDir;
+        expect(await file.exists(out)).toBe(true);
```
```suggestion
        expect(await file.exists(outputDir)).toBe(true);
```

Seems like we don't need to assign the higher-scoped variable to a local const since we use it for an assertion.

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+        time.toMs("1m"),
+      );
+
+      const getDepConImg = (deploy: kind.Deployment, container: string): string => {
```
`DepCon` means DeploymentContainer?

btlghrants replied:
getDepConImg == "get Deployment Container Image" and is just a throw-away name for a throw-away lambda that we use here to avoid having to type out this YAML-object accessor-nightmare out the couple/few times we need it:
```
deploy!.spec!.template!.spec!.containers.filter(f => f.name === container).at(0)!.image!;
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+      it("--timeout, works", async () => {
+        const moduleYaml = await resource.manyFromFile(`${outputDir}/pepr-module-${uuid}.yaml`);
+        {
+          const mwc = resource.select(
```
It's clear that `mwc` means `MutatingWebhookConfiguration` and spelling the whole thing out is a bit much. No issue with this abbreviation

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+            kind.MutatingWebhookConfiguration,
+            `pepr-${uuid}`,
+          );
+          const webhook = mwc.webhooks!.filter(f => f.name === `pepr-${uuid}.pepr.dev`).at(0)!;
```
```suggestion
          const webhook = mwc.webhooks!.filter(file => file.name === `pepr-${uuid}.pepr.dev`).at(0)!;
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+            kind.ValidatingWebhookConfiguration,
+            `pepr-${uuid}`,
+          );
+          const webhook = mwc.webhooks!.filter(f => f.name === `pepr-${uuid}.pepr.dev`).at(0)!;
```
```suggestion
          const webhook = mwc.webhooks!.filter(file => file.name === `pepr-${uuid}.pepr.dev`).at(0)!;
```

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+      });
+
+      it("--withPullSecret, works", async () => {
+        const getDepImgPull = (deploy: kind.Deployment): string[] => {
```
`getDepImgPull` means `getDeploymentImagePull`?

btlghrants replied:
More or less, yes.

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+
+      it("--withPullSecret, works", async () => {
+        const getDepImgPull = (deploy: kind.Deployment): string[] => {
+          return deploy!.spec!.template!.spec!.imagePullSecrets!.map(m => m.name!);
```
What does `m` stand for in this filter? `module`?

btlghrants replied:
"m" stands for "mapped item" (or whatever you'd like) at it is, admittedly, a throwaway identifier (a lot like "x" in any random algebraic function).

The thing that's being mapped there are `.imagePullSecrets` (right there on the same line, immediately before it), so if you were looking for a fully expanded name it'd be something like "imagePullSecret".

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+
+        // TODO: team talk
+        // Image pull secrets don't seem to map into the Helm chart anywhere..? Should they?
+        // TODO: end
```
Let's write this up as an issue and remove the TODO

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+
+      // it("--rbac-mode, ...todo!", async () => {
+      // --rbac-mode scoped
+      // }
```
```suggestion
```

Dangling comment, let's pull it

samayer12 commented on integration/cli/build.nonconflict.test.ts:
```diff
@@ -0,0 +1,187 @@
+
+        const zarfYaml = await resource.oneFromFile(`${outputDir}/zarf.yaml`);
+        const component = zarfYaml.components
+          .filter((f: { name: string }) => f.name === "module")
```
```suggestion
          .filter((file: { name: string }) => file.name === "module")
```

Does `f` represent a `file` here?

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+
+  describe("builds a module", () => {
+    const id = FILE.split(".").at(1);
+    const mod = `${workdir.path()}/${id}`;
```
```suggestion
    const module = `${workdir.path()}/${id}`;
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+    let uuid: string;
+
+    beforeAll(async () => {
+      await fs.rm(mod, { recursive: true, force: true });
```
```suggestion
      await fs.rm(module, { recursive: true, force: true });
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
+      await pepr.cli(mod, { cmd: `npm install` });
```
```suggestion
      await pepr.tgzifyModule(module);
      await pepr.cli(module, { cmd: `npm install` });
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+      const registryInfo = "registry.io/username";
+
+      const getDepConImg = (deploy: kind.Deployment, container: string): string => {
+        return deploy!.spec!.template!.spec!.containers.filter(f => f.name === container).at(0)!
```
```suggestion
        return deploy!.spec!.template!.spec!.containers.filter(file => file.name === container).at(0)!
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+    describe("using a custom registry", () => {
+      const registryInfo = "registry.io/username";
+
+      const getDepConImg = (deploy: kind.Deployment, container: string): string => {
```
This stands for `getDeploymentContainerImage`, yeah?

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+        "builds",
+        async () => {
+          const argz = [`--registry-info ${registryInfo}`].join(" ");
+          const build = await pepr.cli(mod, { cmd: `pepr build ${argz}` });
```
```suggestion
          const build = await pepr.cli(module, { cmd: `pepr build ${argz}` });
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+      );
+
+      it("outputs appropriate configuration", async () => {
+        // overwrites --custom-image..?
```
```suggestion
```

Can we remove this comment? Was it implemented?

btlghrants replied:
It's a pepr CLI design question -- one that I still don't know the answer to.  If the answer is yes then this should result in an update to the Pepr docs explaining that the `--registry-info` flag overwrites the `custom-image` contribution, because it appears to & surprise me here.

btlghrants replied:
That said, I'll remove the comment and add it to the list of follow-on issues.

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+          expect(watcherImage).toBe(image);
+        }
+        {
+          const zarfYaml = await resource.oneFromFile(`${mod}/dist/zarf.yaml`);
```
```suggestion
          const zarfYaml = await resource.oneFromFile(`${module}/dist/zarf.yaml`);
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+          expect(componentImage).toBe(image);
+        }
+        {
+          const valuesYaml = await resource.oneFromFile(`${mod}/dist/${uuid}-chart/values.yaml`);
```
```suggestion
          const valuesYaml = await resource.oneFromFile(`${module}/dist/${uuid}-chart/values.yaml`);
```

samayer12 commented on integration/cli/build.registryinfo.test.ts:
```diff
@@ -0,0 +1,97 @@
+        }
+        {
+          const valuesYaml = await resource.oneFromFile(`${mod}/dist/${uuid}-chart/values.yaml`);
+          const admissionImage = valuesYaml.admission.image;
+          expect(admissionImage).toBe(image);
+
+          const watcherImage = valuesYaml.watcher.image;
+          expect(watcherImage).toBe(image);
```
We expect both images to match `image`. Does that have to do with just how we set up the test scaffolding? Does this match real use?

btlghrants replied:
It does indeed match real use -- that's how that CLI flag functions.

When we pass in a custom image / registry info flag to the `pepr build` command, we have to distribute the single resulting custom image string into 2 separate places within each manifest: one copy for running the Pepr admission controller, and the other for running the Pepr watch controller.

Same image is used for both.

samayer12 commented on integration/cli/build.version.test.ts:
```diff
@@ -0,0 +1,68 @@
+
+  describe("builds a module", () => {
+    const id = FILE.split(".").at(1);
+    const mod = `${workdir.path()}/${id}`;
```
```suggestion
    const module = `${workdir.path()}/${id}`;
```

samayer12 commented on integration/cli/build.version.test.ts:
```diff
@@ -0,0 +1,68 @@
+    const mod = `${workdir.path()}/${id}`;
+
+    beforeAll(async () => {
+      await fs.rm(mod, { recursive: true, force: true });
```
```suggestion
      await fs.rm(module, { recursive: true, force: true });
```

samayer12 commented on integration/cli/build.version.test.ts:
```diff
@@ -0,0 +1,68 @@
+      ].join(" ");
+      await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      await pepr.tgzifyModule(mod);
+      await pepr.cli(mod, { cmd: `npm install` });
```
```suggestion
      await pepr.tgzifyModule(module);
      await pepr.cli(module, { cmd: `npm install` });
```

samayer12 commented on integration/cli/build.version.test.ts:
```diff
@@ -0,0 +1,68 @@
+        "builds",
+        async () => {
+          const argz = [`--version ${version}`].join(" ");
+          const build = await pepr.cli(mod, { cmd: `pepr build ${argz}` });
```
```suggestion
          const build = await pepr.cli(module, { cmd: `pepr build ${argz}` });
```

samayer12 commented on integration/cli/build.version.test.ts:
```diff
@@ -0,0 +1,68 @@
+          // looks like it's just giving back the `pepr --version` then exiting,
+          //  rather than buidling/affecting the output files at all..?
+          expect(await file.exists(`${mod}/dist`)).toBe(false);
+          // TODO: end
```
Let's tag up on this sync and see what the deal is here

btlghrants replied:
OBE -- comment removed.  Issue added to address: https://github.com/defenseunicorns/pepr/issues/1661

Happy to still sync on if you wanna, of course, as part of a different PR though. :+1:

samayer12 commented on integration/cli/build.version.test.ts:
```diff
@@ -0,0 +1,68 @@
+      // it(
+      //   "outputs appropriate configuration",
+      //   async () => { /* TOOD? */ }
+      // );
```
```suggestion
```

Dangling comments

samayer12 commented on integration/cli/init.test.ts:
```diff
@@ -0,0 +1,60 @@
+      const res = await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      expect(res.exitcode).toBe(0);
+      expect(res.stderr.join("").trim()).toBe("");
+      expect(res.stdout.at(0)).toMatch("Usage: pepr init");
```
```suggestion
      const result = await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
      expect(result.exitcode).toBe(0);
      expect(result.stderr.join("").trim()).toBe("");
      expect(result.stdout.at(0)).toMatch("Usage: pepr init");
```

samayer12 commented on integration/cli/init.test.ts:
```diff
@@ -0,0 +1,60 @@
+    async () => {
+      const name = "flags-name";
+      const desc = "flags-desc";
+      const errb = "reject";
```
```suggestion
      const errorBehavior = "reject";
```
`errb` means....? ErrorBehavior?

samayer12 commented on integration/cli/init.test.ts:
```diff
@@ -0,0 +1,60 @@
+      const argz = [
+        `--name ${name}`,
+        `--description ${desc}`,
+        `--errorBehavior ${errb}`,
```
```suggestion
        `--errorBehavior ${errorBehavior}`,
```

samayer12 commented on integration/cli/init.test.ts:
```diff
@@ -0,0 +1,60 @@
+      const res = await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
+      expect(res.exitcode).toBe(0);
+      expect(res.stderr.join("").trim()).toBe("");
+      expect(res.stdout.join("").trim()).toContain("New Pepr module created");
```
```suggestion
      const result = await pepr.cli(workdir.path(), { cmd: `pepr init ${argz}` });
      expect(result.exitcode).toBe(0);
      expect(result.stderr.join("").trim()).toBe("");
      expect(result.stdout.join("").trim()).toContain("New Pepr module created");
```

samayer12 commented on integration/cli/init.test.ts:
```diff
@@ -0,0 +1,60 @@
+      );
+      expect(packageJson.name).toBe(name);
+      expect(packageJson.description).toBe(desc);
+      expect(packageJson.pepr.onError).toBe(errb);
```
```suggestion
      expect(packageJson.pepr.onError).toBe(errorBehavior);
```

samayer12 commented on integration/helpers/cmd.test.ts:
```diff
@@ -0,0 +1,56 @@
+import { describe, expect, it } from "@jest/globals";
+import { Cmd } from "./cmd";
+
+describe("runRaw()", () => {
```
See comments in `cmd.test.ts` for if we actually need this or can use a `DevDependency` instead.

btlghrants replied:
It's what we're already using for running the majority of the e2es assertsion, so... if you want to swap it out with something else, we should definitely talk about that as a separate stream of work (cuz the change won't just be here).

Since this is just the thinnest wrapper around node's native `child_process.spawn` though... until we find we can't make it work / find we need significantly more complex CLI interaction facilities my vote is probably going to be to just leave this as-is.

samayer12 commented on integration/helpers/cmd.ts:
```diff
@@ -0,0 +1,74 @@
+  exitcode: number;
+}
+
+export class Cmd {
```
Why do we prefer our own implementation instead of using something like exec or execSync?

btlghrants replied:
We _definitely_ prefer this, since:
1.  this uses the node `spawn` method to interface with the system (just like we would / did if we chose to call into it more directly
1. this  is already in wide use over in the excellent examples -- [where we had to learn the shortcomings of exec/execSync the hard way](https://github.com/defenseunicorns/pepr-excellent-examples/pull/187) -- and
1. this has clear input/output object semantics which are _way_ easier to read / think through than looking at the buffers directly, having everyone throw together their own wrappers around exec/execSync/spawn wherever the need 'em, etc.

samayer12 commented on integration/helpers/file.ts:
```diff
@@ -0,0 +1,14 @@
+import { access } from "node:fs/promises";
+
+export async function exists(path: string): Promise<boolean> {
```
For tests, it seems like we can afford do these checks synchronously. Can we use [`fs.existsSync()`](https://nodejs.org/api/fs.html#fsexistssyncpath) instead and not make this a concern that we need to maintain and test?

btlghrants replied:
Yes, we can.  I like to stick with the async API when it comes to fs activities (since file access is slow & async APIs mean I don't have to hold up the JS loop to wait) but if we like bouncing between async & sync then the `fs.existsSync` call will work just file.

samayer12 commented on integration/helpers/file.test.ts:
```diff
@@ -0,0 +1,28 @@
+const FILE = path.basename(__filename);
+const HERE = __dirname;
+
+describe("exists", () => {
```
See comments for `file.test.ts`.

btlghrants replied:
Fixed!

samayer12 commented on integration/helpers/workdir.ts:
```diff
@@ -0,0 +1,50 @@
+    return this.path();
+  }
+
+  async exists() {
```
See comment about using `fs.existsSync()` instead. Even if we can't use this, I think it's a duplicate implementation from `file.ts`.

samayer12 commented on integration/prep.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+ME="$(readlink -f "$0")"
```
What does `ME` mean here?

btlghrants replied:
It means "me", literally.  As in, "me, myself, and I".

It contains a fully resolved, absolute path to the script being executed.

samayer12 commented on integration/prep.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+ME="$(readlink -f "$0")"
+HERE="$(dirname "$ME")"
```
Is this similar to the current working-directory?

btlghrants replied:
Nah, not really.  Instead it refers to the directory containing the script being executed.

So, if you're executing the script from _within_ the directory the script sits in, then yes: this value and $(pwd) would return similar paths (though there might still be relative vs absolute issues).

But in the case where you execute this script from a location _outside_ the script directory (say, by running something like `./../this/is/not/in/pwd/script.sh`) then $(pwd) and "$HERE" will be very different.

samayer12 commented on package.json:
```diff
@@ -27,6 +27,9 @@
     "test:unit": "npm run gen-data-json && jest src --coverage --detectOpenHandles --coverageDirectory=./coverage --testPathIgnorePatterns='cosign.e2e.test.ts'",
+    "test:int": "npm run test:int:prep && npm run test:int:run",
+    "test:int:prep": "./integration/prep.sh",
+    "test:int:run": "jest --maxWorkers=4 integration",
```
```suggestion
    "test:integration": "npm run test:int:prep && npm run test:int:run",
    "test:integration:prep": "./integration/prep.sh",
    "test:integration:run": "jest --maxWorkers=4 integration",
```

Unless we see ourselves running the command `npm run test:integration:*` a lot locally, we might as well spell it out if it's mostly used in CI

samayer12 commented on package.json:
```diff
@@ -27,6 +27,9 @@
     "test:unit": "npm run gen-data-json && jest src --coverage --detectOpenHandles --coverageDirectory=./coverage --testPathIgnorePatterns='cosign.e2e.test.ts'",
+    "test:int": "npm run test:int:prep && npm run test:int:run",
+    "test:int:prep": "./integration/prep.sh",
+    "test:int:run": "jest --maxWorkers=4 integration",
```
```suggestion
    "test:integration": "./integration/prep.sh && jest --maxWorkers=4 integration",
```

Alternatively, based on the PR we really only use `test:int` in CI and the nested commands of `prep` and `run` only contain one command each.

So, in interest of avoiding a situation where `A` calls `B` & `C` where `B` calls `1` and `C` calls `2` we can have `A` call `1` and `2` directly. Less bloat for an already chunky `scripts` section.

btlghrants replied:
Nah, would rather not collapse these for dev-flow purposes.

The benefit of the separation is that I can just run the "test:integration:prep" step _once_ to get the pepr-dev-version-.tgz installed and then I can run the "test:integration:run" command a _bunch_ of times as I iterate on the tests.  Saves me 10s-ish every run by not having to run the prep step over and over.

samayer12 commented on .github/workflows/node.js.yml:
```diff
@@ -44,6 +44,25 @@ jobs:
+          version: v3.3.4
+
+      - run: npm ci
+      - run: npm run test:int
```
```suggestion
      - run: npm run test:integration
```

Personal style preference to avoid abbreviations for small words

samayer12 commented on integration/prep.sh:
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
```suggestion
#!/bin/sh
# This script makes a version of the npm cache for local use to avoid installing test artifacts into the global npm cache.
# This isn't an issue in CI where environments are ephemeral, but is useful for local testing.
```

We can add a quick blurb about why we need to prep to get around the npm cache problem.

btlghrants replied:
Good suggestion.

samayer12 commented on integration/helpers/time.ts:
We should avoid taking on time-related utility functions as a concern for Pepr. We should investigate alternatives before concluding this is the best way to represent an arbitrary amount of milliseconds in human-readable formats.

If we accept that adding `devDependencies` for testing is a thing we can do without as much concern as real project dependencies, then we should research some alternatives such as:
* [millis-js](https://www.npmjs.com/package/millis-js)
* [convert-ms](https://www.npmjs.com/package/convert-ms)
* [milliseconds](https://www.npmjs.com/package/milliseconds)
* [convert-hrtime](https://www.npmjs.com/package/convert-hrtime)

btlghrants replied:
I think... I definitely have concerns about pulling dependencies _even if_ they're devDeps (since hacks can happen in CI too).

Open to have a deeper conversation about what position we should take as a project here, but for now this seems a reasonably simple test shim for avoiding random math being done to convert human readable durations into ms-based timeout values.  Shouldn't be much of a problem to rip out & replace later should we find the need to pull in a new lib for dates/times either.

samayer12 commented on integration/helpers/time.test.ts:
```diff
@@ -0,0 +1,64 @@
+import { describe, it, expect } from "@jest/globals";
+import * as sut from "./time";
+
+describe("toHuman()", () => {
```
See comments on `time.ts`. We should avoid taking responsibility for time converting functions unless we have a really good reason to do so.

btlghrants replied:
Don't think of this as "taking responsibility for time",  rather think of this as more of a test helper for human readable durations within tests.

Reason this is here is because `time.toMs("2m30s")` is _infinitely_ more readable than seeing random numbers (i.e. 150000) or math (i.e. `2.5 * 60 * 1000`) sprinkled around everywhere.

I agree about not owning time -- cuz I don't want to have to account for leap years / seconds / time-zones / etc. either -- but this isn't really that; this is more of a reversible, human-readable-duration-to-milliseconds helper function (which we already have in various styles, in various places... just that this is cleaner, and tested).

Alternatives to this are 1) pull in a whole, non-trivial time lib just to do simple formatting (which I don't like for minimizing-dependencies-is-good reasons), 2) remove this and go back to everybody calculating test timeout durations on the fly, or 3) defining local conversion functions wherever we need them (none of which are ideal, IMO).

If, at some point in the future, we find a good reason to pull in a "date/time library" (which I don't really think we have ATM), then I wouldn't object to ripping these out and relying on that instead, but... I'm not convinced that we have grounds for that just yet.

samayer12 replied:
Sure, easy enough to rip out internal calls with a library if we prefer

samayer12 commented on integration/helpers/resource.ts:
Need to review this later, didn't have a chance to look at the details.

samayer12 commented on integration/helpers/resource.test.ts:
Need to review this later, didn't have a chance to look at the details.

samayer12 commented on integration/helpers/pepr.ts:
Need to review this later, didn't have a chance to look at the details.

samayer12 commented on integration/helpers/pepr.test.ts:
Need to review this later, didn't have a chance to look at the details.

btlghrants committed:
wip: PR feedback - swap 'testModule' for 'mod'

btlghrants committed:
wip: PR feedback - swap 'test:integration' for 'test:int'

btlghrants committed:
wip: PR feedback - swap 'test:integration' for 'test:int'

btlghrants committed:
wip: PR feedback - swap 'result' for 'res'

btlghrants committed:
wip: PR feedback - swap 'result' for 'res'

btlghrants committed:
wip: PR feedback - remove 'TODO' comments

samayer12 commented on integration/helpers/resource.ts:
This file appears to let the tester grab data from input that's _either_ JSON or YAML. So, having it as a helper seems reasonable. `manyFromFile()` and `oneFromFile()` seem very similar, I wonder if we can combine them instead of having two implementations that are almost the same.

My mindset is that operating on a collection with `1` item could be the same as operating on a set of `N` items with the right use of calls to things like filter/map/reduce. Maybe.

btlghrants replied:
Yep, good eye.  I thought so too originally.  I tried to combine them early-on but just couldn't get the return types to work out in a way that wouldn't require a caller to have to do a runtime "is this a single or a collection"-style check every time it was used though, so I settled for a two-func api.

Admittedly, it forces the caller state up-front whether they want one or many items back, but... that feels cleaner to me than requiring a caller to do / possibly forget to do an after-call type check.

Definitely open to refinements here if your TS is better than mine.  Minor caveat to watch for -- should you decide to try your hand -- is that the yaml lib we're using has a separate read function for reading from singular vs  multi-doc files, so... there be dagrons.

samayer12 replied:
I'll take a crack at it and report back with suggested changes if I come up with any

samayer12 replied:
@btlghrants, #1663 is a PR to this PR that refactors the parsing logic. The tests in `resource.test.ts` pass locally, but as we noted today there are issues running this workflow in a Mac environment. So, I'm not certain that the integration tests are unaffected but the unit test are g2g.

Merge #1663 into this PR if you're cool with that refactor.

btlghrants replied:
Love it.  Merged it in.

samayer12 commented on integration/helpers/pepr.ts:
```diff
@@ -0,0 +1,50 @@
+
+const HERE = __dirname;
+
+export async function projectRoot(): Promise<string> {
```
I don't think we need to actually export this function since it's only called by `pepr.ts` and `pepr.test.ts`. That probably came about during initial implementation, which makes sense. Based off the rest of the unit tests in `pepr.test.ts`, I think this looks like exposing an implementation-detail of `tgzifyModule()` and `cli()` when do don't need to.

btlghrants replied:
I mean, you _could_ look at it that way, but... I actually intended it to be exposed as it's a totally reasonable thing for other helpers to want to access to / to need to know.

If there's no _problem_ with exposing it, I think I'd prefer to leave it as-is.  Rather than thinking of it as an implementation detail, think of `tgzifyModule` and `cli` as higher-level helpers that build off this as a basic, pepr-module-scoped `projectRoot` helper instead.

samayer12 commented on integration/helpers/pepr.ts:
```diff
@@ -0,0 +1,50 @@
+  await writeFile(packagePath, JSON.stringify(packageJson, null, 2));
+}
+
+export async function copyModule(src: string, dst: string): Promise<void> {
```
`copyModule()` is only called once by `pepr.test.ts` and not used elsewhere. Do we really need it?

btlghrants replied:
Nope -- you're right.  This one was an oversight on my part when I did the conversion from serialized-to-parallelizable test layout I just fogot to remove it.

Consider it gone.

samayer12 reviewed (changes requested):
Had a chance to review `pepr.ts` and `resource.ts`, a few questions about implementation details. The overall purpose of the files makes sense.

btlghrants committed:
wip: PR feedback - remove intermediate local var

btlghrants committed:
Merge branch 'main' into 1640_helm_templates_correspond

btlghrants committed:
wip: PR feedback - expand filter func args

btlghrants committed:
wip: PR feedback - expand map func args

btlghrants committed:
wip: PR feedback - remove placeholder comment

btlghrants committed:
wip: PR feedback - remove flag precedence comment-question

btlghrants committed:
wip: PR feedback - expand short variable name

btlghrants committed:
wip: PR feedback - use fs.existsSync instead of file helper

btlghrants committed:
wip: PR feedback - rm unneeded pepr.copyModule helper

btlghrants committed:
wip: PR feedback - use fs.existsSync in Workdir helper

btlghrants committed:
Update integration/prep.sh

Co-authored-by: Sam Mayer <sam.mayer@defenseunicorns.com>

btlghrants committed:
wip: rm dead import

btlghrants committed:
wip: rm dead import

samayer12 committed:
chore: consolidate file parsing implementation (#1663)

## Description

This PR provides a fix to feedback on #1642 


## Related Issue

Relates to #1640 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging
- [x] Unit,
[Journey](https://github.com/defenseunicorns/pepr/tree/main/journey),
[E2E Tests](https://github.com/defenseunicorns/pepr-excellent-examples),
[docs](https://github.com/defenseunicorns/pepr/tree/main/docs),
[adr](https://github.com/defenseunicorns/pepr/tree/main/adr) added or
updated as needed
- [x] [Contributor Guide
Steps](https://docs.pepr.dev/main/contribute/#submitting-a-pull-request)
followed

btlghrants committed:
chore: Mac support / module func name simplification

btlghrants committed:
chore: merge main

btlghrants committed:
Merge branch 'main' into 1640_helm_templates_correspond

samayer12 reviewed (approved):
Big lift! There are some areas to consolidate and discussions around which helpers we should own vs. list as devdeps, but that's for later.
